File: core/decision.py
"""Decision engine — turn a deterministic valuation into a BUY/WATCH/PASS call
with an evidence trail (Part 1: VALUE, "what's worth buying").

The thesis this module makes concrete: **the math makes the call.** A verdict is
a pure function of three pass/fail criteria; the LLM never votes. The only thing
an agent contributes is the one-sentence `narrative` — colour on top of a decision
that was already made by arithmetic.

Criteria (all deterministic, thresholds tunable via the config's skill block):
  1. Payback Time     < `payback_max_years` (default 12 — a short payback == a
                        cheap business)
  2. Margin of Safety > 0                   (price sits below the fair-value
                        sticker)
  3. Free Cash Flow yield >= `fcf_yield_min` (default 0 == "positive" — the
                        business actually generates cash)

Verdict:
  BUY   = all 3 criteria pass
  WATCH = exactly 2 pass
  PASS  = 1 or 0 pass        ("pass" as in "take a pass on it")

Conviction (0-100) is a smooth, monotonic function of the margin of safety, used
to RANK the universe. Deeper discount == higher conviction == higher rank.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

from .valuation import Valuation

PAYBACK_MAX_YEARS = 12.0
# Default FCF-yield floor for the BUY/WATCH/PASS verdict (0 = "positive").
FCF_YIELD_MIN = 0.0


@dataclass(frozen=True)
class DecisionThresholds:
    """The tunable pass/fail thresholds behind the verdict. Defaults reproduce
    the classic rule (payback < 12y, positive FCF)."""

    payback_max_years: float = PAYBACK_MAX_YEARS
    # FCF-yield floor as a fraction (0.05 = 5%).
    fcf_yield_min: float = FCF_YIELD_MIN


DEFAULT_THRESHOLDS = DecisionThresholds()


def _number(x) -> bool:
    return isinstance(x, (int, float)) and not isinstance(x, bool)


def decision_thresholds(cfg: dict) -> DecisionThresholds:
    """Read the verdict thresholds from a config's skill block, with safe
    defaults (an absent / non-positive `payback_max_years` falls back to 12)."""
    skill = cfg.get("skill", {}) or {}
    pbt = skill.get("payback_max_years")
    fcf = skill.get("fcf_yield_min")
    return DecisionThresholds(
        payback_max_years=pbt if _number(pbt) and pbt > 0 else PAYBACK_MAX_YEARS,
        fcf_yield_min=fcf if _number(fcf) else FCF_YIELD_MIN,
    )


def _fcf_positive(fcf) -> bool:
    """Normalise a config `fcf` field to a boolean. Accepts 'positive'/'negative'
    strings, booleans, or a raw number."""
    if isinstance(fcf, bool):
        return fcf
    if isinstance(fcf, (int, float)):
        return fcf > 0
    if isinstance(fcf, str):
        return fcf.strip().lower() in ("positive", "pos", "+", "true", "yes")
    return False


def conviction_from_mos(mos: float) -> int:
    """Map margin of safety to a 0-100 conviction score.

    A logistic curve centred at mos=0 (conviction 50): a stock exactly at fair
    value is a coin-flip; a deep discount approaches 100; a rich price approaches
    0. Monotonic in `mos`, so ranking by conviction == ranking by discount.
    """
    return int(round(100.0 / (1.0 + math.exp(-mos))))


def valuation_criteria(v: Valuation, fcf_yield: float,
                       thresholds: DecisionThresholds = DEFAULT_THRESHOLDS,
                       ) -> list[dict]:
    """The three-item evidence checklist for a valuation, each with value,
    threshold, and a deterministic pass/fail. `fcf_yield` is a fraction; the
    thresholds default to the classic rule when omitted."""
    pbt = v.payback_years
    mos = v.margin_of_safety
    pbt_max = thresholds.payback_max_years
    fcf_min = thresholds.fcf_yield_min
    return [
        {
            "name": "Payback Time",
            "value": f"{pbt:.1f}y",
            "threshold": f"< {pbt_max:.0f}y",
            "passed": pbt < pbt_max,
        },
        {
            "name": "Margin of Safety",
            "value": f"{mos * 100:+.0f}%",
            "threshold": "> 0",
            "passed": mos > 0.0,
        },
        {
            "name": "Free Cash Flow yield",
            "value": f"{fcf_yield * 100:.1f}%",
            "threshold": (f"≥ {fcf_min * 100:.0f}%" if fcf_min == 0
                          else f"≥ {fcf_min * 100:.1f}%"),
            "passed": fcf_yield >= fcf_min,
        },
    ]


def verdict_from_criteria(criteria: list[dict]) -> str:
    """3 pass -> BUY, exactly 2 -> WATCH, <=1 -> PASS. Deterministic."""
    passed = sum(1 for c in criteria if c["passed"])
    if passed == 3:
        return "BUY"
    if passed == 2:
        return "WATCH"
    return "PASS"


def decide_valuation(v: Valuation, fcf_yield: float, narrative: str = "",
                     thresholds: DecisionThresholds = DEFAULT_THRESHOLDS,
                     ) -> dict:
    """Build the `valuation` decision block for one stock."""
    criteria = valuation_criteria(v, fcf_yield, thresholds)
    return {
        "verdict": verdict_from_criteria(criteria),
        "conviction": conviction_from_mos(v.margin_of_safety),
        "criteria": criteria,
        "narrative": narrative,
        # A few raw numbers the dashboard likes to show alongside the checklist.
        "payback_years": round(v.payback_years, 2),
        "sticker_price": round(v.sticker, 2),
        "buy_price": round(v.buy_price, 2),
        "margin_of_safety": round(v.margin_of_safety, 4),
    }


@dataclass
class Stock:
    """One watchlist row's static facts (everything except computed verdicts)."""

    ticker: str
    name: str
    sector: str
    price: float
    valuation: Valuation
    fcf_positive: bool
    # FCF yield as a fraction — the value the verdict's FCF criterion tests.
    fcf_yield: float
    narrative: str
    shape: str  # timing sample-series shape: floor | neutral | extended


def stocks_from_config(cfg: dict) -> list[Stock]:
    """Parse the config watchlist into Stock objects, reusing the valuation
    defaults + growth-capping adapter from `core.screen` (the valuation trusts
    a capped growth; the row's own `growth_rate` stays the true CAGR)."""
    from .screen import _defaults, valuation_from_row

    d = _defaults(cfg)
    out: list[Stock] = []
    for row in cfg.get("skill", {}).get("watchlist", []):
        out.append(
            Stock(
                ticker=row["ticker"],
                name=row.get("name", row["ticker"]),
                sector=row.get("sector", ""),
                price=row["price"],
                valuation=valuation_from_row(row, d),
                fcf_positive=_fcf_positive(row.get("fcf", "positive")),
                fcf_yield=(row["fcf_yield"]
                           if _number(row.get("fcf_yield")) else 0.0),
                narrative=row.get("narrative", ""),
                shape=row.get("shape", "neutral"),
            )
        )
    return out


def build_decision(stock: Stock, series=None,
                   thresholds: DecisionThresholds = DEFAULT_THRESHOLDS) -> dict:
    """Assemble the full per-stock decision object (both lenses).

    Args:
        stock: parsed watchlist row.
        series: optional (highs, lows, closes) tuple for the timing lens. When
            omitted, the timing block is computed from the stock's sample shape.
        thresholds: verdict thresholds (defaults to the classic rule).
    """
    from . import signals as sig
    from . import sample_prices as sp

    valuation = decide_valuation(stock.valuation, stock.fcf_yield,
                                 stock.narrative, thresholds)

    if series is None:
        highs, lows, closes = sp.sample_ohlc(stock.ticker, stock.shape)
    else:
        highs, lows, closes = series
    timing = sig.timing_signals(closes, highs=highs, lows=lows)

    return {
        "ticker": stock.ticker,
        "name": stock.name,
        "sector": stock.sector,
        "price": round(stock.price, 2),
        "valuation": valuation,
        "timing": timing,
    }


def build_decisions(cfg: dict) -> list[dict]:
    """Full decision objects for the whole watchlist, ranked by conviction desc.

    Tie-break: margin of safety desc, then ticker asc — a total, deterministic
    order so the dashboard cards are stable run to run.
    """
    stocks = stocks_from_config(cfg)
    thresholds = decision_thresholds(cfg)
    decisions = [build_decision(s, thresholds=thresholds) for s in stocks]
    decisions.sort(
        key=lambda d: (
            -d["valuation"]["conviction"],
            -d["valuation"]["margin_of_safety"],
            d["ticker"],
        )
    )
    for i, d in enumerate(decisions, start=1):
        d["rank"] = i
    return decisions

File: web/engine/decision.ts
/**
 * Decision engine — turn a deterministic valuation into a BUY/WATCH/PASS call
 * with an evidence trail. TypeScript port of core/decision.py.
 *
 * The thesis this module makes concrete: **the math makes the call.** A verdict
 * is a pure function of three pass/fail criteria; the LLM never votes.
 *
 * Criteria (all deterministic):
 *   1. Payback Time    < 12 years   (a short payback == a cheap business)
 *   2. Margin of Safety > 0         (price sits below the fair-value sticker)
 *   3. Free Cash Flow   positive    (the business actually generates cash)
 *
 * Verdict: BUY = all 3 pass · WATCH = exactly 2 · PASS = 1 or 0.
 * Conviction (0-100) is a smooth, monotonic function of the margin of safety.
 */

import { pyFixed, pyRound } from "./pyformat.ts";
import { Valuation } from "./valuation.ts";

export const PAYBACK_MAX_YEARS = 12.0;
/** Default FCF-yield floor for the BUY/WATCH/PASS verdict (0 = "positive"). */
export const FCF_YIELD_MIN = 0.0;

export type Verdict = "BUY" | "WATCH" | "PASS";

/**
 * The tunable pass/fail thresholds behind the verdict. Defaults reproduce the
 * classic rule (payback < 12y, positive FCF); the terminal's Advanced settings
 * let a visitor move them.
 */
export interface DecisionThresholds {
  paybackMaxYears: number;
  /** FCF-yield floor as a fraction (0.05 = 5%). */
  fcfYieldMin: number;
}

export const DEFAULT_THRESHOLDS: DecisionThresholds = {
  paybackMaxYears: PAYBACK_MAX_YEARS,
  fcfYieldMin: FCF_YIELD_MIN,
};

/** Read the verdict thresholds from a config's skill block, with safe defaults. */
export function decisionThresholds(cfg: Config): DecisionThresholds {
  const skill = cfg.skill ?? {};
  const pbt = skill.payback_max_years;
  const fcf = skill.fcf_yield_min;
  return {
    paybackMaxYears: typeof pbt === "number" && pbt > 0 ? pbt : PAYBACK_MAX_YEARS,
    fcfYieldMin: typeof fcf === "number" ? fcf : FCF_YIELD_MIN,
  };
}

/**
 * Normalise a config `fcf` field to a boolean. Accepts 'positive'/'negative'
 * strings, booleans, or a raw number.
 */
export function fcfPositive(fcf: unknown): boolean {
  if (typeof fcf === "boolean") return fcf;
  if (typeof fcf === "number") return fcf > 0;
  if (typeof fcf === "string") {
    return ["positive", "pos", "+", "true", "yes"].includes(fcf.trim().toLowerCase());
  }
  return false;
}

/**
 * Map margin of safety to a 0-100 conviction score: a logistic curve centred
 * at mos=0 (conviction 50). Monotonic, so ranking by conviction == ranking by
 * discount.
 */
export function convictionFromMos(mos: number): number {
  return pyRound(100.0 / (1.0 + Math.exp(-mos)), 0);
}

export interface Criterion {
  name: string;
  value: string;
  threshold: string;
  passed: boolean;
}

/**
 * The three-item evidence checklist for a valuation, each with value,
 * threshold, and a deterministic pass/fail. `fcfYield` is a fraction; the
 * thresholds default to the classic rule when omitted.
 */
export function valuationCriteria(
  v: Valuation, fcfYield: number, thresholds: DecisionThresholds = DEFAULT_THRESHOLDS,
): Criterion[] {
  const pbt = v.paybackYears;
  const mos = v.marginOfSafety;
  const { paybackMaxYears, fcfYieldMin } = thresholds;
  return [
    {
      name: "Payback Time",
      value: `${pyFixed(pbt, 1)}y`,
      threshold: `< ${pyFixed(paybackMaxYears, 0)}y`,
      passed: pbt < paybackMaxYears,
    },
    {
      name: "Margin of Safety",
      value: `${pyFixed(mos * 100, 0, true)}%`,
      threshold: "> 0",
      passed: mos > 0.0,
    },
    {
      name: "Free Cash Flow yield",
      value: `${pyFixed(fcfYield * 100, 1)}%`,
      threshold: `≥ ${pyFixed(fcfYieldMin * 100, fcfYieldMin === 0 ? 0 : 1)}%`,
      passed: fcfYield >= fcfYieldMin,
    },
  ];
}

/** 3 pass -> BUY, exactly 2 -> WATCH, <=1 -> PASS. Deterministic. */
export function verdictFromCriteria(criteria: readonly Criterion[]): Verdict {
  const passed = criteria.reduce((acc, c) => acc + (c.passed ? 1 : 0), 0);
  if (passed === 3) return "BUY";
  if (passed === 2) return "WATCH";
  return "PASS";
}

export interface ValuationDecision {
  verdict: Verdict;
  conviction: number;
  criteria: Criterion[];
  narrative: string;
  payback_years: number;
  sticker_price: number;
  buy_price: number;
  margin_of_safety: number;
}

/** Build the `valuation` decision block for one stock. */
export function decideValuation(
  v: Valuation, fcfYield: number, narrative: string = "",
  thresholds: DecisionThresholds = DEFAULT_THRESHOLDS,
): ValuationDecision {
  const criteria = valuationCriteria(v, fcfYield, thresholds);
  return {
    verdict: verdictFromCriteria(criteria),
    conviction: convictionFromMos(v.marginOfSafety),
    criteria,
    narrative,
    // A few raw numbers the dashboard likes to show alongside the checklist.
    payback_years: pyRound(v.paybackYears, 2),
    sticker_price: pyRound(v.sticker, 2),
    buy_price: pyRound(v.buyPrice, 2),
    margin_of_safety: pyRound(v.marginOfSafety, 4),
  };
}

// ---- watchlist parsing ------------------------------------------------------

/** One config watchlist row (config.example.json / config.json shape). */
export interface ConfigRow {
  ticker: string;
  name?: string;
  sector?: string;
  price: number;
  eps: number;
  growth_rate: number;
  future_pe?: number;
  fcf?: unknown;
  fcf_yield?: number;
  shape?: string;
  narrative?: string;
  [key: string]: unknown;
}

export interface Config {
  skill?: {
    projection_years?: number;
    required_return?: number;
    margin_of_safety?: number;
    default_future_pe?: number;
    /** Ceiling on the growth the valuation trusts (display growth is untouched). */
    valuation_growth_cap?: number;
    /** Verdict thresholds (watchlist BUY/WATCH/PASS). */
    payback_max_years?: number;
    /** FCF-yield floor as a fraction (0.05 = 5%). */
    fcf_yield_min?: number;
    watchlist?: ConfigRow[];
    [key: string]: unknown;
  };
  [key: string]: unknown;
}

/** One watchlist row's static facts (everything except computed verdicts). */
export interface Stock {
  ticker: string;
  name: string;
  sector: string;
  price: number;
  valuation: Valuation;
  fcfPositive: boolean;
  /** FCF yield as a fraction — the value the verdict's FCF criterion tests. */
  fcfYield: number;
  narrative: string;
  /** timing sample-series shape: floor | neutral | extended */
  shape: string;
}

export interface ValuationDefaults {
  years: number;
  requiredReturn: number;
  margin: number;
  defaultFuturePe: number;
  /** Ceiling on the growth the valuation trusts (Infinity = uncapped). */
  valuationGrowthCap: number;
}

/** Pull global valuation assumptions from the config's skill block. */
export function configDefaults(cfg: Config): ValuationDefaults {
  const skill = cfg.skill ?? {};
  const cap = skill.valuation_growth_cap;
  return {
    years: skill.projection_years ?? 10,
    requiredReturn: skill.required_return ?? 0.15,
    margin: skill.margin_of_safety ?? 0.50,
    defaultFuturePe: skill.default_future_pe ?? 15.0,
    // A non-finite / non-positive cap means "don't cap" (uncapped valuation).
    valuationGrowthCap: typeof cap === "number" && cap > 0 ? cap : Infinity,
  };
}

/**
 * Build a Valuation for one config row using the global defaults. The growth
 * the valuation trusts is capped at `valuationGrowthCap`; the row's own
 * `growth_rate` (the true historical CAGR shown in the watchlist) is untouched.
 */
export function valuationFromRow(row: ConfigRow, d: ValuationDefaults): Valuation {
  return new Valuation({
    ticker: row.ticker,
    price: row.price,
    eps: row.eps,
    growthRate: Math.min(row.growth_rate, d.valuationGrowthCap),
    futurePe: row.future_pe ?? d.defaultFuturePe,
    years: d.years,
    requiredReturn: d.requiredReturn,
    margin: d.margin,
  });
}

/** Parse the config watchlist into Stock objects. */
export function stocksFromConfig(cfg: Config): Stock[] {
  const d = configDefaults(cfg);
  const rows = cfg.skill?.watchlist ?? [];
  return rows.map((row) => ({
    ticker: row.ticker,
    name: row.name ?? row.ticker,
    sector: row.sector ?? "",
    price: row.price,
    valuation: valuationFromRow(row, d),
    fcfPositive: fcfPositive(row.fcf ?? "positive"),
    fcfYield: typeof row.fcf_yield === "number" ? row.fcf_yield : 0,
    narrative: row.narrative ?? "",
    shape: row.shape ?? "neutral",
  }));
}
